src/smp/ipi.c

```diff
@@ -21,62 +21,69 @@
-         * Note that, 'ThreadState_Restart' does not always result in regenerating exception
-         * if we are in kernel to handle them, e.g. hardware single step exception. */
-        if (thread_state_ptr_get_tsType(&NODE_STATE(ksCurThread)->tcbState) == ThreadState_Running) {
-            setThreadState(NODE_STATE(ksCurThread), ThreadState_Restart);
-        }
-
+    // ipiStallCore can only be called for a single core at a time to be able to assume
+    // kernel mutual exclusion as the caller node is blocked until this stall completes
+    // and all other nodes are out of the kernel or blocked on the lock.
+    assert(totalCoreBarrier == 1);
+
+    // There's 3 ways that this function can be called:
+    // - On the IRQ path where the lock isn't acquired
+    // - On the IRQ path where the lock is attempted to be acquired
+    // - On the non-IRQ path where the lock is attempted to be acquired
+    // !irqPath implies clh_is_self_in_queue()
+    assert(clh_is_self_in_queue() || irqPath);
+
+    /* The current thread is running as we would replace this thread with an idle thread
+     *
+     * The instruction should be re-executed if we are in kernel to handle syscalls.
+     * Also, thread in 'ThreadState_RunningVM' should remain in same state.
+     * Note that, 'ThreadState_Restart' does not always result in regenerating exception
+     * if we are in kernel to handle them, e.g. hardware single step exception. */
+    if (!irqPath && thread_state_ptr_get_tsType(&NODE_STATE(ksCurThread)->tcbState) == ThreadState_Running) {
+        setThreadState(NODE_STATE(ksCurThread), ThreadState_Restart);
+    }
+#ifdef CONFIG_KERNEL_MCS
+    updateTimestamp();
+#endif
+    if (config_ternary(CONFIG_KERNEL_MCS, checkBudget(),1)) {
         SCHED_ENQUEUE_CURRENT_TCB;
-        switchToIdleThread();
+    }
+    switchToIdleThread();
+    NODE_STATE(ksSchedulerAction) = SchedulerAction_ResumeCurrentThread;
 #ifdef CONFIG_KERNEL_MCS
-        commitTime();
-        NODE_STATE(ksCurSC) = NODE_STATE(ksIdleThread)->tcbSchedContext;
+    switchSchedContext();
+    NODE_STATE(ksReprogram) = false;
 #endif
-        NODE_STATE(ksSchedulerAction) = SchedulerAction_ResumeCurrentThread;
 
-        /* Let the cpu requesting this IPI to continue while we waiting on lock */
-        big_kernel_lock.node_owners[getCurrentCPUIndex()].ipi = 0;
+    if (irqPath) {
+        return;
+    }
+
+    /* Let the cpu requesting this IPI to continue while we waiting on lock */
+    big_kernel_lock.node_owners[getCurrentCPUIndex()].ipi = 0;
 #ifdef CONFIG_ARCH_RISCV
-        ipi_clear_irq(irq_remote_call_ipi);
+    ipi_clear_irq(irq_remote_call_ipi);
 #endif
-        ipi_wait(totalCoreBarrier);
+    ipi_wait(totalCoreBarrier);
 
-        /* Continue waiting on lock */
-        while (big_kernel_lock.node_owners[getCurrentCPUIndex()].next->value != CLHState_Granted) {
-            if (clh_is_ipi_pending(getCurrentCPUIndex())) {
+    /* Continue waiting on lock */
```

Comment:
I think I accidentally reverted a comment I had here:
- Because there's no way to remove a node from the lock queue until it has acquired the lock and can then release it, in the syscall context, we need to wait for the lock here and then immediately return.  We can't return to higher up the stack because the syscall context is no longer valid since the thread was changed.